wpilib/src/robot.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright 2019 First Rust Competition Developers.
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use super::{
+    ds::{DriverStation, RobotState},
+    notifier::Alarm,
+    RobotBase,
+};
+use std::time;
+use wpilib_sys::{
+    HAL_ObserveUserProgramAutonomous, HAL_ObserveUserProgramDisabled,
+    HAL_ObserveUserProgramStarting, HAL_ObserveUserProgramTeleop, HAL_ObserveUserProgramTest,
+};
+
+/// Implements a specific type of robot program framework, for
+/// `start_iterative` and `start_timed`.
+///
+/// The init methods are called whenever the appropriate mode is entered.
+///
+/// The periodic functions are called for the appropriate mode on an interval.
+pub trait IterativeRobot {
+    fn disabled_init(&mut self) {
+        println!("Default disabled_init method... Override me!");
+    }
+    fn autonomous_init(&mut self) {
+        println!("Default autonomous_init method... Override me!");
+    }
+    fn teleop_init(&mut self) {
+        println!("Default teleop_init method... Override me!");
+    }
+    fn test_init(&mut self) {
+        println!("Default test_init method... Override me!");
+    }
+
+    /// Periodic code for all modes should go here.
+    fn robot_periodic(&mut self) {}
+
+    fn disabled_periodic(&mut self) {}
+    fn autonomous_periodic(&mut self) {}
+    fn teleop_periodic(&mut self) {}
+    fn test_periodic(&mut self) {}
+}
+
+fn loop_func<T: IterativeRobot>(
+    robot: &mut T,
+    last_mode: Option<RobotState>,
+    cur_mode: RobotState,
+) {
+    // Check for state transitions
+    match (last_mode, cur_mode) {
```

Comment:
There are a lot of unused matches here that could probably be reduced.
```rust
if let Some(last_mode) == last_mode {
    if last_mode != cur_mode {
        
        match cur_mode {
            RobotState::Disabled => robot.disabled_init(),
            RobotState::Autonomous => robot.autonomous_init(),
            // etc.
        }
    }
}
```

Reply:
We need to also match on None. But sure, we could refactor out the `init` calls.

Reply:
Wait, of course, Option<T> implements PartialEq for T: PartialEq and RobotState derives PartialEq. Cleaning this up now.